crates/services/producer/src/block_producer.rs

```diff
@@ -88,6 +95,65 @@ pub struct Producer<ViewProvider, TxPool, Executor, GasPriceProvider, ConsensusP
     pub consensus_parameters_provider: ConsensusProvider,
 }
 
+impl<ViewProvider, TxPool, Executor, GasPriceProvider, ConsensusProvider>
+    Producer<ViewProvider, TxPool, Executor, GasPriceProvider, ConsensusProvider>
+where
+    ViewProvider: AtomicView + 'static,
+    ViewProvider::LatestView: BlockProducerDatabase,
+    ConsensusProvider: ConsensusParametersProvider,
+{
+    pub async fn produce_and_execute_predefined(
+        &self,
+        predefined_block: &Block,
+    ) -> anyhow::Result<UncommittedResult<Changes>>
+    where
+        Executor: ports::BlockProducer<Vec<Transaction>> + 'static,
+    {
+        let _production_guard = self.lock.lock().await;
+
+        let mut transactions_source = predefined_block.transactions().to_vec();
+
+        let height = predefined_block.header().consensus().height;
+
+        let block_time = predefined_block.header().consensus().time;
+
+        let da_height = predefined_block.header().application().da_height;
+
+        let header_to_produce = self
+            .new_header_with_da_height(height, block_time, da_height)
+            .await?;
+
+        let maybe_mint_tx = transactions_source.pop();
+        let mint_tx =
+            maybe_mint_tx
+                .and_then(|tx| tx.as_mint().cloned())
+                .ok_or(anyhow!(
+                    "The last transaction in the block should be a mint transaction"
+                ))?;
+
+        let gas_price = *mint_tx.gas_price();
+        let coinbase_recipient = mint_tx.input_contract().contract_id;
+
+        let component = Components {
+            header_to_produce,
+            transactions_source,
+            coinbase_recipient,
+            gas_price,
+        };
+
+        // Store the context string in case we error.
+        let context_string =
+            format!("Failed to produce block {height:?} due to execution failure");
+        let result = self
+            .executor
+            .produce_without_commit(component)
+            .map_err(Into::<anyhow::Error>::into)
+            .context(context_string)?;
```

Comment:
nit: This should probably use [`with_context`](https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.with_context) instead, although the allocation on happy path is likely optimized away anyway.